Framework/API/src/MatrixWorkspace.cpp

```diff
@@ -83,6 +84,31 @@ MatrixWorkspace::~MatrixWorkspace() {
   }
 }
 
+void MatrixWorkspace::setIndexTranslator(
+    Indexing::SpectrumNumbers &&spectrumNumbers,
+    Indexing::DetectorIDs &&detectorIDs) {
+  setIndexTranslator(Indexing::IndexTranslator(std::move(spectrumNumbers),
+                                               std::move(detectorIDs)));
+}
+
+void MatrixWorkspace::setIndexTranslator(
+    Indexing::IndexTranslator &&translator) {
+  if (translator.size() != getNumberHistograms())
+    throw std::runtime_error("MatrixWorkspace::setIndexTranslator: Translator "
+                             "size does not match number of histograms in "
+                             "workspace");
+
+  auto &spectrumNumbers = translator.spectrumNumbers();
+  auto &detectorIDs = translator.detectorIDs();
+  for (size_t i = 0; i < getNumberHistograms(); ++i) {
+    auto &spectrum = getSpectrum(i);
+    spectrum.setSpectrumNo(spectrumNumbers[i]);
+    if (!detectorIDs.empty())
```

Comment:
@OwenArnold Example for the first: Remember the `Indexing` prototyping I did, there was a separate `SpectrumNumberTranslator` and `DetectorIDTranslator`. We could then have

```cpp
MatrixWorkspace::setSpectrumNumberTranslator(...);
MatrixWorkspace::setDetectorIDTranslator(...);

// update spectrum numbers in workspace
ws.setSpectrumNumberTranslator(SpectrumNumbers{1,2,3,4,5});

// update detector IDs in workspace
ws.setDetectorIDs(DetectorIDs{100,101,102,103,105});
```

Basically there would be two independent setters, and probably also two independent translator objects.